numba/np/arraymath.py

```diff
@@ -355,117 +354,133 @@ def array_sum_impl_axis(arr, axis):
 @overload(np.prod)
 @overload_method(types.Array, "prod")
 def array_prod(arr):
+    if isinstance(arr, types.Array):
+        def array_prod_impl(arr):
+            c = 1
+            for v in np.nditer(arr):
+                c *= v.item()
+            return c
 
-    def array_prod_impl(arr):
-        c = 1
-        for v in np.nditer(arr):
-            c *= v.item()
-        return c
-
-    return array_prod_impl
+        return array_prod_impl
 
 
 @overload(np.cumsum)
 @overload_method(types.Array, "cumsum")
 def array_cumsum(arr):
-    is_integer = arr.dtype in types.signed_domain | frozenset([types.bool_])
+    if isinstance(arr, types.Array):
+        is_integer = arr.dtype in types.signed_domain | frozenset([types.bool_])
 
-    def array_cumsum_impl(arr):
-        if is_integer:
-            dtype = np.intp
-        else:
-            dtype = arr.dtype
-        out = np.empty(arr.size, dtype)
-        c = np.zeros(1, dtype)[0]
-        for idx, v in enumerate(arr.flat):
-            c += v
-            out[idx] = c
-        return out
+        def array_cumsum_impl(arr):
+            if is_integer:
+                dtype = np.intp
+            else:
+                dtype = arr.dtype
+            out = np.empty(arr.size, dtype)
+            c = np.zeros(1, dtype)[0]
+            for idx, v in enumerate(arr.flat):
+                c += v
+                out[idx] = c
+            return out
 
-    return array_cumsum_impl
+        return array_cumsum_impl
 
 
 @overload(np.cumprod)
 @overload_method(types.Array, "cumprod")
 def array_cumprod(arr):
-    is_integer = arr.dtype in types.signed_domain | frozenset([types.bool_])
+    if isinstance(arr, types.Array):
+        is_integer = arr.dtype in types.signed_domain | frozenset([types.bool_])
 
-    def array_cumprod_impl(arr):
-        if is_integer:
-            dtype = np.intp
-        else:
-            dtype = arr.dtype
-        out = np.empty(arr.size, dtype)
-        c = np.ones(1, dtype)[0]
-        for idx, v in enumerate(arr.flat):
-            c *= v
-            out[idx] = c
-        return out
+        def array_cumprod_impl(arr):
+            if is_integer:
+                dtype = np.intp
+            else:
+                dtype = arr.dtype
```

Comment:
Same here! You can move the dtype assignment to outside the impl. function.

Reply:
Resolved in [4e38728](https://github.com/numba/numba/pull/8415/commits/4e38728908f9f1583ebd37e4b694f68d9f4c5c97)